gnuradio-runtime/lib/top_block_impl.cc

```diff
@@ -129,6 +132,21 @@ namespace gr {
   void
   top_block_impl::wait()
   {
+    do {
```

Comment:
Just to add proper boundaries for lock. The d_lock_count  must be examined under lock, but wait_for_jobs() must be in unlocked section of the code.

This is just loop, but is loopend through the 'continue' , not trought the condition on while().

I still have the feeling I did not get what You want to now.